[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51368 [PyTorch] Use real if constexpr behind macro in hot template
* #51367 [PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor
* **#51319 [PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor**
* #51463 [PyTorch] Fix return value of IValue::to for Tensor/String
* #51315 [PyTorch][easy] Fix TODOs in CppFunction constructors
* #51313 [PyTorch] Reduce template expansion in call_functor_with_args_from_stack

We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

swolchok committed:
[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor

We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:
```diff
@@ -249,29 +249,68 @@ namespace impl {
   template<class T, bool AllowDeprecatedTypes>
   struct ivalue_to_arg final {
-    static T call(IValue&& v) {
+    static decltype(auto) call(IValue& v) {
```
curious, what does this change do?

swolchok replied:
It makes the return type match the declared type of the expression we return on the following line, rather than always returning by value. https://en.cppreference.com/w/cpp/language/function#Return_type_deduction

bhosmer reviewed (commented):
Haven't stared long enough to understand what's provoking the build errors, are you seeing them locally?

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

swolchok committed:
fix op_registration_test build on "[PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor"


We were going out of our way to accommodate `IValue::to<Tensor>` returning a copy of the inner Tensor. `IValue::toTensor` is capable of returning a reference without copying, so if we use it directly, we can allow kernels that want to take `Tensor &` to do so!
As a bonus, we get reduced build times.

Differential Revision: [D26138549](https://our.internmc.facebook.com/intern/diff/D26138549/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -639,7 +639,7 @@ TEST(OperatorRegistrationTest, givenLambdaKernel_whenRegisteringWithMismatchingC
     auto registrar = c10::RegisterOperators().op("_test::dummy", c10::RegisterOperators::options()
       .kernel(DispatchKey::CPU, [] (const int64_t&) {})
-      .kernel(DispatchKey::CUDA, [] (int64_t&) {}));
+      .kernel(DispatchKey::CUDA, [] (int64_t) {}));
```
I'm not completely following why this change is needed but no need to explain unless it'll be useful to future readers of the PR 😁

swolchok commented:
You can't bind an lvalue reference to a prvalue (the returned value from ivalue_to_arg). This code only worked because of reference_cast.

Sent from my iPhone

On Feb 10, 2021, at 5:03 PM, Basil Hosmer <notifications@github.com> wrote:

﻿

@bhosmer commented on this pull request.

________________________________

In aten/src/ATen/core/op_registration/op_registration_test.cpp<https://github.com/pytorch/pytorch/pull/51319#discussion_r574192254>:

> @@ -639,7 +639,7 @@ TEST(OperatorRegistrationTest, givenLambdaKernel_whenRegisteringWithMismatchingC
   expectThrows<c10::Error>([] {
     auto registrar = c10::RegisterOperators().op("_test::dummy", c10::RegisterOperators::options()
       .kernel(DispatchKey::CPU, [] (const int64_t&) {})
-      .kernel(DispatchKey::CUDA, [] (int64_t&) {}));
+      .kernel(DispatchKey::CUDA, [] (int64_t) {}));


I'm not completely following why this change is needed but no need to explain unless it'll be useful to future readers of the PR 😁

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/51319#pullrequestreview-588188068>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAAIK6FXH325YBIZ2QON2CLS6MUGRANCNFSM4WX62FOA>.
